### DISCUSSION TITLE
Support for yolov4 darknet

### DISCUSSION THREAD

**Discussion by neeraj-satyaki** [miguelaeh/pipeless#124] [Q&A] [answered]
Hi, Great SDk for inferencing on Video streams parllely, do you guys support for Yolov4 Darknet weights too?

**Comment by miguelaeh** [comment]
Hi @neeraj-satyaki,

You can load any model and use any framework. You can either use the ONNX Runtime (so you will provide pre and post process functions only in your stage) or even use Pytorch, Tensorflow, or anything else by creating a process function (and also pre and post process).

This is an example of the first, directly loading a model in onnx format: https://github.com/pipeless-ai/pipeless/tree/main/examples/onnx-candy
This is an example of the second, which uses the Ultralytics SDK for YOLOv8: https://github.com/pipeless-ai/pipeless/tree/main/examples/yolo

**Reply by neeraj-satyaki** [reply]
Do we have to add anything specific in preprocess.py

**Reply by miguelaeh** [reply]
No, you don't even have to create that file if not required. You tipically provide it when your model requires a specific image format, size, etc so you perform the modifications there and assign it to inference input. If your model can receive the raw RGB image with the same size as it comes from the source you do not need to even create a pre-process file

**Reply by neeraj-satyaki** [reply]
My Process.json is as follows
{
    "runtime": "onnx",
    "model_uri": "file:///models/yolov4.onnx",
    "inference_params": {
        "execution_provider": "TensorRT"
    }
}
![image](https://github.com/pipeless-ai/pipeless/assets/83659541/ef57e734-ac4e-43b2-82ea-5ab6b6061469)
As you can see in the image, I am unable to run this on Tensorrt or its not picking up GPU 
[2024-01-21T16:27:14Z INFO  ort::memory] new_cpu; allocator=Arena memory_type=Default
[2024-01-21T16:27:14Z INFO  ort::memory] drop; self=MemoryInfo { ptr: 0x7f0a5c928b60, should_release: true }
[2024-01-21T16:27:14Z INFO  ort::memory] new; allocation_device=CPU device_id=0 allocator_type=Device memory_type=Default
[2024-01-21T16:27:14Z INFO  ort::memory] drop; self=MemoryInfo { ptr: 0x7f0a5c642be0, should_release: true }
[2024-01-21T16:27:14Z INFO  ort::memory] new; allocation_device=CPU device_id=0 allocator_type=Device memory_type=Default
[2024-01-21T16:27:14Z INFO  ort::memory] drop; self=MemoryInfo { ptr: 0x7f0a5c932020, should_release: true }
[2024-01-21T16:27:14Z INFO  ort::memory] new_cpu; allocator=Arena memory_type=Default
[2024-01-21T16:27:14Z INFO  ort::memory] drop; self=MemoryInfo { ptr: 0x7f0a5c929d20, should_release: true }
[2024-01-21T16:27:14Z INFO  ort::memory] new; allocation_device=CPU device_id=0 allocator_type=Device memory_type=Default
[2024-01-21T16:27:14Z INFO  ort::memory] drop; self=MemoryInfo { ptr: 0x7f0a5c644620, should_release: true }
[2024-01-21T16:27:14Z INFO  ort::memory] new; allocation_device=CPU device_id=0 allocator_type=Device memory_type=Default
[2024-01-21T16:27:14Z INFO  ort::memory] drop; self=MemoryInfo { ptr: 0x7f0a5c641fa0, should_release: true }
[2024-01-21T16:27:15Z INFO  ort::memory] new_cpu; allocator=Arena memory_type=Default
[2024-01-21T16:27:15Z INFO  ort::memory] drop; self=MemoryInfo { ptr: 0x7f0a5c63c1a0, should_release: true }
[2024-01-21T16:27:15Z INFO  ort::memory] new; allocation_device=CPU device_id=0 allocator_type=Device memory_type=Default
[2024-01-21T16:27:15Z INFO  ort::memory] drop; self=MemoryInfo { ptr: 0x7f0a5c63cfe0, should_release: true }
[2024-01-21T16:27:15Z INFO  ort::memory] new; allocation_device=CPU device_id=0 allocator_type=Device memory_type=Default
[2024-01-21T16:27:15Z INFO  ort::memory] drop; self=MemoryInfo { ptr: 0x7f0a5c6408e0, should_release: true }
[2024-01-21T16:27:15Z INFO  ort::memory] new_cpu; allocator=Arena memory_type=Default
[2024-01-21T16:27:15Z INFO  ort::memory] drop; self=MemoryInfo { ptr: 0x7f0a5c63eee0, should_release: true }
[2024-01-21T16:27:15Z INFO  ort::memory] new; allocation_device=CPU device_id=0 allocator_type=Device memory_type=Default
[2024-01-21T16:27:15Z INFO  ort::memory] drop; self=MemoryInfo { ptr: 0x7f0a5c92a8a0, should_release: true }
[2024-01-21T16:27:15Z INFO  ort::memory] new; allocation_device=CPU device_id=0 allocator_type=Device memory_type=Default
[2024-01-21T16:27:15Z INFO  ort::memory] drop; self=MemoryInfo { ptr: 0x7f0a5c92cb60, should_release: true }
[2024-01-21T16:27:15Z INFO  ort::memory] new_cpu; allocator=Arena memory_type=Default
[2024-01-21T16:27:15Z INFO  ort::memory] drop; self=MemoryInfo { ptr: 0x7f0a5c537ba0, should_release: true }
[2024-01-21T16:27:15Z INFO  ort::memory] new; allocation_device=CPU device_id=0 allocator_type=Device memory_type=Default
[2024-01-21T16:27:15Z INFO  ort::memory] drop; self=MemoryInfo { ptr: 0x7f0a5c92d0a0, should_release: true }
[2024-01-21T16:27:15Z INFO  ort::memory] new; allocation_device=CPU device_id=0 allocator_type=Device memory_type=Default
[2024-01-21T16:27:15Z INFO  ort::memory] drop; self=MemoryInfo { ptr: 0x7f0a5c92dba0, should_release: true }

These are the logs above,
Below is the command,
docker run --gpus all -v ${PWD}/:/app -v /home/ubuntu/Models:/models -e "PIPELESS_USER_PYTHON_PACKAGES=opencv-python;ultralytics;onnxruntime;tensorrt;" -p 3030:3030 miguelaeh/pipeless start --stages-dir /app

Could please help me out on how to use

**Reply by miguelaeh** [reply, marked as answer]
Hi @neeraj-satyaki !

(This is just a productivity advice): It will be more comfortable for you to work if you also mount a volume into `/.venv` because in that way you will avoid installing the Python packages on every docker run, so you will work faster. See this section: https://www.pipeless.ai/docs/docs/v1/container#install-custom-python-packages

Regarding your issue, I just checked it and it seems you are right, there is an issue in the TensorRT container image due to some packages not being found by`libonnxruntime_providers_tensorrt.so`. If you install Pipeless in your host directly it will work fine.
I am going to create a new PR fixing the image now and will link it here. Thanks for the report!

**Reply by miguelaeh** [reply]
@neeraj-satyaki I forgot to tell you your command is also using an image only valid for CPU. You have to use `miguelaeh/pipeless:latest-tensorrt` if you use `latest` it will be CPU only. At the beggining of this page you can see the complete list of images: https://www.pipeless.ai/docs/docs/v1/container

And, this PR should fix the problem: https://github.com/pipeless-ai/pipeless/pull/125

**Reply by neeraj-satyaki** [reply]
![image](https://github.com/pipeless-ai/pipeless/assets/83659541/f3fd901e-b2a3-44ae-b53b-ef902cac7566)
I was uptill here, for ```miguelaeh/pipeless:latest-tensorrt```. Could Tell me where to move through.
Sorry For asking you too much. Just Felt Stuck so

**Reply by miguelaeh** [reply]
Ey @neeraj-satyaki no problem!

The first issue you see about TensorRT failing to load is the same error we commented yesterday. I merged the PR, however, I am having trouble to release the new image in [this job](https://github.com/pipeless-ai/pipeless/actions/runs/7613859842/job/20734782168), for some reason it gets stuck (seems to be a known error of the docker buildx github action).

Then you have an error that is probably related to a wrong file path to `yolov8n-seg-psda.pt`. It seems you are using a file with that name in a `init.py` file, right? Ensure the path to the file is absolute and remember it must be a valid path inside the container not in your host.

The third error (the one stated by the parser when loading stages) seems to me like you mounted the `/.venv` folder inside `/app`. It must be mounted like `-v "/tmp/venv:/.venv"` (i.e. at `/.venv`). The Pipeless project (`/app` directory) cannot contain files that Pipeless do not expect (yet) so it is failing when parsing that unexpected venv folder at `/app/.venv`. Ensure you mount it at `/.venv` instead.

**Reply by neeraj-satyaki** [reply]
Thank you so much, for your insights. For now can i use 
Tensorrt version of the image

**Reply by miguelaeh** [reply]
You should be able to use the image in a few hours. I am going to create a new release now and once the workflows complete the container with tag `v1.6.3-tensorrt` (that will become latest-tensorrt once released) should be fine.

**Reply by miguelaeh** [reply]
@neeraj-satyaki the image is already published, you need to pull it again `docker pull miguelaeh/pipeless:latest-tensorrt` or `docker pull miguelaeh/pipeless:v1.6.3-tensorrt`

**Reply by neeraj-satyaki** [reply]
Hi @miguelaeh, Faced new issue, I did get it running on GPu,
![image](https://github.com/pipeless-ai/pipeless/assets/83659541/58a8bcd3-d3b8-4b3b-934f-956029d297ac)
My Pre and Post Process hooks are
def hook(frame_data, _):
    frame = frame_data['original'].astype('float32') 
    print("Inout Frame, ", frame.shape)
    frame_data['inference_input'] = frame
    
def hook(frame_data, _):
    frame = frame_data['modified'].astype('float32') 
    print("masks", frame.shape)
    frame_data['modified'] = frame
    
I just wanted to verify what would be the output coming in.
Did I miss something while Passing the model or the image?

**Reply by miguelaeh** [reply]
That seems to be related to how you are providing the RGB image to the model. The model you loaded expects the channels as the first element like (3, 480,640) and it seems you are providing the inference inut as (480,640,3). You probably need to permute the array colums before assigning it to the inference input in your pre processing.

**Reply by neeraj-satyaki** [reply]
Ok, got it.
